File: Deniz/day_8.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 is None:
            return list2
        if list2 is None:
            return list1

        prehead = ListNode(-1)
        
        prev = prehead

        p1 = list1
        p2 = list2

        while p1 is not None and p2 is not None:
            if p1.val <= p2.val:
                prev.next = p1
                p1 = p1.next
            else:
                prev.next = p2
                p2 = p2.next
            prev = prev.next 

        if p1 is not None:
            prev.next = p1
        elif p2 is not None:
            prev.next = p2

        return prehead.next

File: Murat/day_8.cpp
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* head = nullptr; 
        ListNode** cur = &head; // cursor
        
        while (list1 || list2) {
            if (list1 && (!list2 || list1->val < list2->val)) {
                *cur = list1; // Point current node to list1 node
                list1 = list1->next;
            } else if (list2) {
                *cur = list2; // Point current node to list2 node
                list2 = list2->next; 
            }
            cur = &(*cur)->next;
        }
        
        return head;
    }
};
